### PR TITLE
optimized reserve call for memory allocation in advance

### PR DESCRIPTION
**Pull request by blacroc10**
may help reduce re allocations during concatenation

### CODE REVIEW

File: libsmtutil/SMTLib2Interface.cpp [hunk 1]
```diff
@@ -189,29 +189,37 @@ std::string SMTLib2Interface::toSExpr(solidity::smtutil::Expression const& _expr
...
-
-	return command;
+    std::string command;
+    if (_expressionsToEvaluate.empty())
+        command = "(check-sat)\n";
+    else
+    {
+        // Reserve space for efficiency
+        size_t numExpressions = _expressionsToEvaluate.size();
+        command.reserve(command.size() + numExpressions * 100); // Estimate space needed
```
Comment (nikola-matic) [inline libsmtutil/SMTLib2Interface.cpp:199]: Where did you come up with the 100 chars per expression heuristic? Also, you're not using `shrink_to_fit()` before returning the command, so you're most likely returning a much larger command string than you should be.

@blishko I'd be inclined towards closing this is a mostly unnecessary optimization, but it's up to you.

Comment (blishko) [reply libsmtutil/SMTLib2Interface.cpp:199]: I agree that this is unnecessary optimisation, with a magic number involved.
If anything, `command` should be a `stringstream`. That would also achieve the desired result in a clearer way.
But this is definitely not any bottleneck, AFAIK.

Comment (cameel) [reply libsmtutil/SMTLib2Interface.cpp:199]: Caching of `vector`'s `size()` is unnecessary as well. It's guaranteed O(1) and introducing an extra local variable for something so trivial is a hit to readability (it's another name you have to keep track of in your head when reading the code) so not a good trade-off.

I'd not be against us using things like `reserve()` more in places where the size is non-trivial and known ahead of time, but TBH most of the time the impact is probably insignificant so it does not matter all that much in the big picture. Just changing it randomly in a single spot that is not a known bottleneck is pointless.

**Comment by nikola-matic** [comment]
Given the current sentiment regarding this PR, I'll close it as 'won't do'. Regardless of this not being merged, thanks for contributing @blacroc10!

**Comment by cameel** [comment]
@blacroc10 Sorry, but this is the kind of contribution we're generally not interested in. The PR is just a random micro-optimization for the sake of it. It's also very sloppy (e.g. it completely disregards the style of the surrounding code by rewriting randomly replacing all indents in the middle of a file).

**Pull request closed without merging**